nvflare/apis/impl/client_controller.py

```diff
@@ -0,0 +1,223 @@
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.security.logging import secure_format_exception
+
+
+class ClientController(FLComponent, ControllerSpec):
+    def __init__(
+        self,
+    ) -> None:
+        super().__init__()
+        self.task_data_filters = None
+        self.task_result_filters = None
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.START_RUN:
+            self.start_controller(fl_ctx)
+        elif event_type == EventType.END_RUN:
+            self.stop_controller(fl_ctx)
+
+    def start_controller(self, fl_ctx: FLContext):
+        client_runner = fl_ctx.get_prop(FLContextKey.RUNNER)
+        self.task_data_filters = client_runner.task_data_filters
+        self.task_result_filters = client_runner.task_result_filters
+
+    def stop_controller(self, fl_ctx: FLContext):
+        pass
+
+    def process_result_of_unknown_task(
+        self, client: Client, task_name: str, client_task_id: str, result: Shareable, fl_ctx: FLContext
+    ):
+        pass
+
+    def broadcast(
+        self,
+        task: Task,
+        fl_ctx: FLContext,
+        targets: Union[List[Client], List[str], None] = None,
+        min_responses: int = 0,
+        wait_time_after_min_received: int = 0,
+    ):
+
+        return self.broadcast_and_wait(task, fl_ctx, targets, min_responses, wait_time_after_min_received)
+
+    def broadcast_and_wait(
+        self,
+        task: Task,
+        fl_ctx: FLContext,
+        targets: Union[List[Client], List[str], None] = None,
+        min_responses: int = 0,
+        wait_time_after_min_received: int = 0,
+        abort_signal: Signal = None,
+    ):
+        engine = fl_ctx.get_engine()
+        request = task.data
+        # apply task filters
+        self.log_debug(fl_ctx, "firing event EventType.BEFORE_TASK_DATA_FILTER")
+        self.fire_event(EventType.BEFORE_TASK_DATA_FILTER, fl_ctx)
+
+        # # first apply privacy-defined filters
+        task_filter_list = self.task_data_filters.get(task.name)
+        filter_error, task.data = self.apply_data_filters(task_filter_list, request, fl_ctx)
+
+        if filter_error:
+            replies = self._make_error_reply(ReturnCode.TASK_DATA_FILTER_ERROR, targets)
+            return replies
+
+        # set up ClientTask for each client
+        for target in targets:
+            client: Client = self._get_client(target, engine)
+            client_task = ClientTask(task=task, client=client)
+            task.client_tasks.append(client_task)
+            task.last_client_task_map[client_task.id] = client_task
+
+            # task_cb_error = self._call_task_cb(task.before_task_sent_cb, client, task, fl_ctx)
+            # if task_cb_error:
+            #     return self._make_error_reply(ReturnCode.ERROR, targets)
```

Comment:
Does this mean for this PR we don't support before_task_sent_cb in client side controller?

Reply:
That's correct. There's no reason (no need) to use before_task_sent_cb when creating Task in client controller.